catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/DescribableServiceMap.java

```diff
@@ -0,0 +1,137 @@
+/**
+ * Copyright (c) Codice Foundation
+ * <p>
+ * This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
+ * General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ * <p>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package ddf.catalog.util.impl;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ddf.catalog.util.Describable;
+
+public class DescribableServiceMap<V extends Describable> implements Map<String, V> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DescribableServiceMap.class);
+
+    private static final String READ_ONLY_ERROR_MESSAGE = "This map is meant to be read only.";
+
+    private Map<ServiceReference, Describable> serviceRefMap =
+            Collections.synchronizedMap(new HashMap<>());
+
+    private Map<String, V> serviceMap = Collections.synchronizedMap(new HashMap<>());
+
+    public DescribableServiceMap() {
+
+    }
+
+    protected BundleContext getContext() {
+        Bundle cxfBundle = FrameworkUtil.getBundle(SortedServiceList.class);
+        if (cxfBundle != null) {
+            return cxfBundle.getBundleContext();
+        }
+        return null;
+    }
+
+    public void bind(ServiceReference ref) {
+
+        LOGGER.debug(this + " Binding " + ref);
+        BundleContext context = getContext();
+
+        if (context != null) {
+            V service = (V) context.getService(ref);
+            serviceRefMap.put(ref, service);
+            serviceMap.put(service.getId(), service);
+        } else {
+            LOGGER.debug("BundleContext was null, unable to add service reference");
+        }
+    }
+
+    public void unbind(ServiceReference ref) {
+
+        LOGGER.debug("Unbinding " + ref);
+
+        Describable service = serviceRefMap.remove(ref);
```

Comment:
It seems weird to me to hang on to both maps with the one map only used to figure out access the other map. Isn't there any other way to get from the ServiceReference to the id that you need? Is it in the properties or something?